### PR TITLE
support zstd levels

### PR DESCRIPTION
**Pull request by popsikle**
We are running into bandwidth issues and want to be able to configure zstd compression levels.

### CODE REVIEW

**Review by jonhyman — APPROVED** [review]

File: lib/mongo/protocol/compressed.rb [hunk 1]
```diff
@@ -72,17 +72,17 @@ class Compressed < Message
...
       # @since 2.5.0
-      def initialize(message, compressor, zlib_compression_level = nil)
+      def initialize(message, compressor, compression_level = nil)
         @original_message = message
         @original_op_code = message.op_code
         @uncompressed_size = 0
         @compressor_id = COMPRESSOR_ID_MAP[compressor]
         @compressed_message = ''
-        @zlib_compression_level = zlib_compression_level if zlib_compression_level && zlib_compression_level != -1
+        @compression_level = compression_level if compression_level && compression_level != -1
```
Comment (jasonpenny) [inline lib/mongo/protocol/compressed.rb:85]: zstd allows negative compression levels

File: lib/mongo/protocol/message.rb [hunk 2]
```diff
@@ -102,14 +102,14 @@ def replyable?
       # the command being sent permits compression. Otherwise returns self.
       #
       # @param [ String, Symbol ] compressor The compressor to use.
-      # @param [ Integer ] zlib_compression_level The zlib compression level to use.
+      # @param [ Integer ] compression_level The zlib compression level to use.
```
Comment (jasonpenny) [inline lib/mongo/protocol/message.rb:105]:
```suggestion
      # @param [ Integer ] compression_level The compression level to use.
```

File: lib/mongo/protocol/message.rb [hunk 3]
```diff
@@ -118,15 +118,15 @@ def maybe_compress(compressor, zlib_compression_level = nil)
       #
       # @param [ String ] command_name Command name extracted from the message.
       # @param [ String | Symbol ] compressor The compressor to use.
-      # @param [ Integer ] zlib_compression_level Zlib compression level to use.
+      # @param [ Integer ] compression_level Zlib compression level to use.
```
Comment (jasonpenny) [inline lib/mongo/protocol/message.rb:121]:
```suggestion
      # @param [ Integer ] compression_level Compression level to use.
```

File: lib/mongo/protocol/msg.rb [hunk 4]
```diff
@@ -161,15 +161,15 @@ def serialize(buffer = BSON::ByteBuffer.new, max_bson_size = nil, bson_overhead
       # Otherwise returns self.
       #
       # @param [ String, Symbol ] compressor The compressor to use.
-      # @param [ Integer ] zlib_compression_level The zlib compression level to use.
+      # @param [ Integer ] compression_level The zlib compression level to use.
```
Comment (jasonpenny) [inline lib/mongo/protocol/msg.rb:164]:
```suggestion
      # @param [ Integer ] compression_level The compression level to use.
```

File: lib/mongo/uri/options_mapper.rb [hunk 5]
```diff
@@ -594,6 +595,31 @@ def revert_zlib_compression_level(value)
...
+      def convert_zstd_compression_level(name, value)
+        if /\A-?\d+\z/ =~ value
+          i = value.to_i
+
+          if i >= 1 && i <= 22
+            return i
+          end
+        end
+
+        log_warn("#{value} is not a valid zstdCompressionLevel")
```
Comment (jasonpenny) [inline lib/mongo/uri/options_mapper.rb:614]: negative compression levels are supported by zstd

https://github.com/SpringMT/zstd-ruby/blob/d682a90d84a54137309574f2b007ab11ecf02f71/ext/zstdruby/libzstd/zstd.h#L81-L85

**Pull request merged**